### PR TITLE
Don't `sort` chain by default

### PR DESCRIPTION
**Pull request by torfjelde**
This drops the default behavior of sorting the chain in `sample`. Sorting can be turned on again using the kwarg `sort_chain=true`.

IMO we should get https://github.com/TuringLang/DynamicPPL.jl/pull/251 merged first, so that uses of `setval!` and others are working correctly. This requires upping the compat-entry for DPPL, and so I figured we might as well also add re-export of `@submodel` in this PR.

### CODE REVIEW

File: src/Turing.jl [hunk 1]
```diff
@@ -82,6 +82,7 @@ using StatsBase: predict
 # Turing essentials - modelling macros and inference algorithms
 export  @model,                 # modelling
         @varname,
+        @submodel,
```
Comment (devmotion) [inline src/Turing.jl:85]: Is this intended? I'm fine with it but it seems a bit unrelated here :slightly_smiling_face:

File: src/inference/Inference.jl [hunk 2]
```diff
@@ -356,13 +357,15 @@ function AbstractMCMC.bundle_samples(
...
+    chain = MCMCChains.Chains(
         parray,
         nms,
         (internals = extra_params,);
         evidence=le,
         info=info,
-    ) |> sort
+    )
+
+    return sort_chain ? sort(chain) : chain
```
Comment (devmotion) [inline src/inference/Inference.jl:368]: Ah I think I remember why it was removed - doesn't this create a type instability? IIRC `sample` can't be inferred due to limitations of Julia's logging system anyway, so I guess it doesn't matter :shrug:

File: src/Turing.jl (hunk 1 above)
Comment (cpfiffer) [reply src/Turing.jl:85]: Yeah I saw someone on Slack noted that it wasn't exported. I think it's fine to bundle in here.

**Comment by cpfiffer** [comment]
We should probably go up a minor version since this is a big change from the UI perspective.

File: src/Turing.jl (hunk 1 above)
Comment (torfjelde) [reply src/Turing.jl:85]: Yep:) From above:

> This requires upping the compat-entry for DPPL, and so I figured we might as well also add re-export of @submodel in this PR.

Comment (devmotion) [reply src/Turing.jl:85]: Ah missed your comment :+1:

File: src/inference/Inference.jl (hunk 2 above)
Comment (torfjelde) [reply src/inference/Inference.jl:368]: Ah, but do we care? :confused:

Comment (devmotion) [reply src/inference/Inference.jl:368]: As I said, I am pretty sure it's unstable either way, so one should always use a function barrier if it's important. Hence I'm fine with the change :slightly_smiling_face:

**Review by devmotion — COMMENTED** [review]
Looks good, maybe add a test as well?

**Comment by torfjelde** [comment]
> Looks good, maybe add a test as well?

Will do :+1: Just waiting for the DPPL release and I'll add tests and get this through.

File: test/inference/Inference.jl [hunk 3]
```diff
@@ -131,4 +131,12 @@
         @test mean(x[:s][1] for x in chains) ≈ 3 atol=0.1
         @test mean(x[:m][1] for x in chains) ≈ 0 atol=0.1
     end
+
+    @testset "chain ordering" begin
+        chain_sorted = sample(gdemo_default, Prior(), 1, sort_chain=true)
+        @test names(MCMCChains.get_sections(chain_sorted, :parameters)) == [:m, :s]
+
+        chain_unsorted = sample(gdemo_default, Prior(), 1, sort_chain=false)
+        @test names(MCMCChains.get_sections(chain_unsorted, :parameters)) == [:m, :s]
```
Comment (devmotion) [inline test/inference/Inference.jl:140]: Not completely surprising, this test fails. I guess it should be
```suggestion
        @test names(MCMCChains.get_sections(chain_unsorted, :parameters)) == [:s, :m]
```
?

Comment (devmotion) [reply test/inference/Inference.jl:140]: BTW it might be good to check the output without keyword argument as well, to check that the default value of `sort_chains` is set as expected.

Comment (devmotion) [reply test/inference/Inference.jl:140]: One additional thought: maybe `sort_parameters` would be more explicit since the chain is a 3D array and hence in principle could be sorted with respect to different axes?

Comment (torfjelde) [reply test/inference/Inference.jl:140]: Ehh yes:)

**Comment by yebai** [comment]
nice work - thanks, @torfjelde and @devmotion!
